File: LLM_experiments/python/results_gpt-4.1/solution_try_1/needleman_wunsch.py
import json
from typing import Dict


def load_sequences(filepath: str) -> Dict[str, str]:
    with open(filepath, "r", encoding="utf-8") as f:
        data = json.load(f)
    return data


def load_blosum(filepath: str) -> Dict[str, int]:
    with open(filepath, "r", encoding="utf-8") as f:
        data = json.load(f)
    return {k: int(v) for k, v in data.items()}


def needleman_wunsch(
    seq1: str, seq2: str, blosum: Dict[str, int]
) -> int:
    m, n = len(seq1), len(seq2)
    score = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        score[i][0] = score[i - 1][0] + blosum[seq1[i - 1]]

    for j in range(1, n + 1):
        score[0][j] = score[0][j - 1] + blosum[seq2[j - 1]]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            char1 = seq1[i - 1]
            char2 = seq2[j - 1]
            match_key = char1 + char2
            match = score[i - 1][j - 1] + blosum[match_key]
            delete = score[i - 1][j] + blosum[char1]
            insert = score[i][j - 1] + blosum[char2]
            score[i][j] = max(match, delete, insert)

    return int(score[m][n])

File: LLM_experiments/java/results_gpt-4.1/solution_try_1/src/main/java/com/example/phylo/NeedlemanWunsch.java
package com.example.phylo;

public class NeedlemanWunsch {
  public static int align(String seq1, String seq2, BlosumMatrix matrix) {
    int m = seq1.length();
    int n = seq2.length();
    int[][] dp = new int[m + 1][n + 1];

    dp[0][0] = 0;
    for (int i = 1; i <= m; i++) {
      dp[i][0] = dp[i - 1][0] + matrix.getDeletionScore(seq1.charAt(i - 1));
    }
    for (int j = 1; j <= n; j++) {
      dp[0][j] = dp[0][j - 1] + matrix.getDeletionScore(seq2.charAt(j - 1));
    }

    for (int i = 1; i <= m; i++) {
      for (int j = 1; j <= n; j++) {
        int match = dp[i - 1][j - 1] + matrix.getSubstitutionScore(seq1.charAt(i - 1), seq2.charAt(j - 1));
        int delete = dp[i - 1][j] + matrix.getDeletionScore(seq1.charAt(i - 1));
        int insert = dp[i][j - 1] + matrix.getDeletionScore(seq2.charAt(j - 1));
        dp[i][j] = Math.max(match, Math.max(delete, insert));
      }
    }
    return dp[m][n];
  }
}